### PR TITLE
fix(ts): include leading operator in `TSUnionType` and `TSIntersectionType` locations

### LINKED ISSUE #12754: @babel/parser(ts): TSUnionType is not including leading pipe into range
## Bug Report

- [ ] I would like to work on a fix!

**Current behavior**

`TSUnionType` with leading pipe `|` has is not including leading pipe character into its range.

this issue is also present for `TSIntersectionType`

- [REPL babel](https://astexplorer.net/#/gist/7ea7e76e3a444af545b8ed14aca37181/6d8d93d31e30eafe2081e34c33d484f18b34ab2c)
- [REPL typescript](https://astexplorer.net/#/gist/7ea7e76e3a444af545b8ed14aca37181/c4cb2430bc58fa52b1c25738dc2af746764c1b10)
- [REPL typescript-eslint](https://astexplorer.net/#/gist/7ea7e76e3a444af545b8ed14aca37181/049f5f728616d2afd4ccec455e21fdddfe3f5b94)

**Input Code**

```ts
type x = 
  | string
  | number
```

**Expected behavior**
Babel should include leading `|` character into range of TSUnionType, 
note: leading `|` is optional and can be omitted

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
enabled plugins

```js
[
    'classProperties',
    'decorators-legacy',
    'estree',
    'typescript',
]
```

**Environment**

- Babel version(s): 7.12.14
- Node/npm version: 10-15
- OS: win/linux/mac
- Monorepo: no
- How you are using Babel: parse(....)

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by fedeci**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12754 
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

We save the location of the leading operator token and use that to start `TSUnionType` and `TSIntersectionType` nodes. 

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12757"></a>

### CODE REVIEW

File: packages/babel-parser/test/fixtures/typescript/type-alias/union-intersection/input.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+type J = number | string
```
Comment (existentialism) [inline packages/babel-parser/test/fixtures/typescript/type-alias/union-intersection/input.ts:1]: Can we just augment:

https://github.com/babel/babel/blob/main/packages/babel-parser/test/fixtures/typescript/types/union-intersection/input.ts

instead?

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 2]
```diff
@@ -935,16 +935,21 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       parseConstituentType: () => N.TsType,
       operator: TokenType,
     ): N.TsType {
-      this.eat(operator);
+      const leadingOperator = this.eat(operator);
+      const leadingOperatorTokStartLoc = this.state.lastTokStartLoc;
+      const leadingOperatorTokStart = this.state.lastTokStart;
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:940]:
```suggestion
      const node: N.TsUnionType | N.TsIntersectionType = this.startNode();
      const leadingOperator = this.eat(operator);
```
We can always start node earlier.

Comment (armano2) [reply packages/babel-parser/src/plugins/typescript/index.js:940]: there is one more case that should be taken into account:
```
type M = | number
```

typescript in this case generates union type to and babel does not

**Comment by armano2** [comment]
there is one more case that should be taken into account:
```ts
type M = | number
type M = & number
```

typescript in this case generates union type to and babel does not, unless this should be done in separate PR

**Comment by JLHwung** [comment]
`typescript-eslint` also generates `TSUnionType`: https://astexplorer.net/#/gist/26a69b81ec7db30e30ec05b387673f6a/f69200243fd8bdff13b85238e53688490c0adae0

I think we should align to TypeScript and `typescript-eslint`. @fedeci Can you work on a fix on that?

**Comment by fedeci** [comment]
@JLHwung yes, should I open a new PR for that?

**Comment by existentialism** [comment]
@fedeci yeah let's do a new pr

**Review by existentialism — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Thanks!

**Pull request merged**